src/app/app_execution.rs

```diff
@@ -108,3 +110,179 @@ impl AppExecution {
         Ok(())
     }
 }
+
+/// A stream of [`RecordBatch`]es that can be paginated for display in the TUI.
+pub struct PaginatingRecordBatchStream {
+    // currently executing stream
+    inner: SendableRecordBatchStream,
+    // any batches that have been buffered so far
+    batches: Vec<RecordBatch>,
+    // current batch being shown
+    current_batch: Option<usize>,
+}
+
+impl PaginatingRecordBatchStream {
```

Comment:
I'm not sure yet if this is the final api signatures - I will know once I plug into the display code